base/workerpool.jl

```diff
@@ -107,10 +124,170 @@ end
+function on the worker nodes, especially when the same function or closure is called repeatedly.
+This is particularly true for closures which capture large amounts of data.
+
+The remote cache is maintained for the lifetime of the returned `CachingPool` object.
+
+Global variables are not captured by a closure. Use `remoteset!` on a `CachingPool` to broadcast
+global variables or global constants.
+"""
+function CachingPool(workers::Vector{Int})
+    pool = CachingPool()
+    for w in workers
+        push!(pool, w)
+    end
+    return pool
+end
+
+CachingPool(wp::WorkerPool) = CachingPool(workers(wp))
+
+function empty!(pool::CachingPool)
+    for (_,rr) in pool.map_obj2ref
+        finalize(rr)
+    end
+    empty!(pool.map_obj2ref)
+
+    results = asyncmap(p->remotecall_fetch(clear_definitions, p, pool.broadcasted_symbols), workers(pool))
+    @assert all(results)
+    empty!(pool.broadcasted_symbols)
+    return pool
+end
+
+cp_clear(pool::CachingPool) = (@schedule empty!(pool); pool)
+
+function exec_from_cache(f, rr::RemoteChannel, args...; kwargs...)
+    if f === nothing
+        if !isready(rr)
+            error("Requested function not found in cache")
+        else
+            fetch(rr)(args...; kwargs...)
+        end
+    else
+        if isready(rr)
+            warn("Serialized function found in cache. Removing from cache.")
+            take!(rr)
+        end
+        put!(rr, f)
+        f(args...; kwargs...)
+    end
+end
+
+function remotecall_pool(rc_f, f, pool::CachingPool, args...; kwargs...)
+    worker = take!(pool)
+    if haskey(pool.map_obj2ref, (worker, f))
+        rr = pool.map_obj2ref[(worker, f)]
+        remote_f = nothing
+    else
+        # TODO : Handle map_obj2ref state if `f` is unable to be cached remotely.
+        rr = RemoteChannel(worker)
+        pool.map_obj2ref[(worker, f)] = rr
+        remote_f = f
+    end
+
+    try
+        rc_f(exec_from_cache, worker, remote_f, rr, args...; kwargs...)
+    finally
+        worker in pool.workers && put!(pool, worker)
+    end
+end
+
+function define_locally(isconst::Bool, mod::Symbol, kvpairs)
+    mod = getmod(mod)
+
+    for (k,v) in kvpairs
+        if isconst
+            eval(mod, Expr(:const, Expr(:(=), k, v)))
+        else
+            eval(mod, Expr(:(=), k, v))
+        end
+    end
+    return true
+end
+
+function clear_definitions(names)
+    for (mod, nm) in names
+        eval(mod, Expr(:(=), nm, nothing))
+    end
+    return true
+end
+
+"""
+    remoteset!(pool::CachingPool, isconst::Bool=true, mod=pool.default_bcast_mod; kwargs...)
+
+Provides a means of defining and setting variables on all workers in the pool.
+Variables to be defined are passed as keyword args with the name as the arg name and value
+as the arg value.
+
+This function should only be used to broadcast variables that will last the entire duration
+of a program run as global variables once declared cannot be unbound.
```

Comment:
"entire duration of a program, since global variables cannot be unbound once declared."